sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidLogicalValuesRule.java

```diff
@@ -110,6 +111,13 @@ static Object getValueFromLiteral(RexLiteral literal, PlannerContext plannerCont
       case SMALLINT:
       case INTEGER:
       case BIGINT:
+        // Safegaurd in case the internal implementaion of the RexLiteral for numbers change
+        Object maybeBigDecimalImpl = literal.getValue();
+        if (maybeBigDecimalImpl instanceof BigDecimal) {
+          return ((BigDecimal) maybeBigDecimalImpl).longValue();
+        }
+        // This might return incorrect value if the literal was created from float.
+        // For example, representation of the form 123.0 can return 1230
```

Comment:
BTW, I was not sure if this is a bug or not, so posted a question in the calcite dev (https://lists.apache.org/thread/yr26vot133q9p7kjjh8nlq6pb1mjx9vl).

Reply:
Per Julian's suggestion in the linked thread, this case statement should be simply `return literal.getValueAs(BigDecimal.class).longValue();`

Reply:
given that all of our calls to `getValueAs` are already within a case statement on the literal type, it seems redundant to me to use `getValueAs`, which is going to do another case statement on typeName instead of type.getTypeName.

Maybe we should just do `literal.getValue` instead of `literal.getValueAs`. Strings I think we can just pass through directly, numeric types will always be `BigDecimal`, so we can just convert to float, double, or long as appropriate.

Reply:
this makes more assumptions about how that code is written inside the library. Calling `literal.getValueAs(BigDecimal.class)` makes the intention explicit that `BigDecimal` is the type we want. Current ways seem future-proof to unexpected changes in the calcite library.

Reply:
I mean we sort of already have to know how the library works to call `literal.getValueAs(BigDecimal.class)` to get a type we don't want to turn it into one we do so it doesn't really seem any different to me. This is the only place we are using `getValueAs` in our code, most other callers use `RexLiteral.value(...)` which is equivalent to `getValue` for `RexLiteral` but also allows `CAST` and unary negation functions to get literal values. Most of those callers appear to be casting the value to `Number` (see `Expressions.literalToDruidExpression`)

Reply:
> I mean we sort of already have to know how the library works to call literal.getValueAs(BigDecimal.class)

That's true. Though we can still try to minimize relying on the internal details. This code path is only called during planning (AFAIK) so there isn't much performance advantage. So the redundancy doesn't cause any real problem anyway.

Reply:
>This code path is only called during planning (AFAIK) so there isn't much performance advantage. So the redundancy doesn't cause any real problem anyway.

Ah, I wasn't really thinking on the performance angle (though planning does a lot of duplicate work), more on consistency with the way the rest of Druid interacts with literals - like we aren't using `getValueAs` anywhere else, so mostly wondering why we should use it here when it isn't really that useful to what we are doing. Using `RexLiteral.stringValue(literal)` and `((Number) RexLiteral.value(literal)` would probably be most consistent I think?

Reply:
I used `getValueAs()` before because it seemed easier than implementing casts in this method. I don't worry about consistency much but, if we decide to refactor this method, I agree that it is important to have the least dependency on Calcite's internal logic. In this sense, using `RexLiteral.value()` and `RexLiteral.stringValue()` seems the best as Clint suggested since it returns the raw `value`. However, refactoring doesn't seem to have a big impact at least for now. So, I don't have a strong opinion here and am OK with just fixing the bug without refactoring.